lib/AST/ASTDumper.cpp

```diff
@@ -2297,6 +2420,28 @@ namespace {
       printFoot();
     }
 
+    // Prints a mapping from the declared interface types of the opaque types to
+    // their equivalent existential type. This loses some information, but it is
+    // meant to make it easier to determine which protocols an opaque type
+    // conforms to when such a type appears elsewhere in an expression dump,
+    // farther away from where the opaque type is declared.
+    void printOpaqueTypeMapping(ArrayRef<OpaqueTypeDecl *> opaqueDecls) {
+      printRecArbitrary(
+          [&](Label label) {
+            printHead("opaque_to_existential_mapping", FieldLabelColor, label);
+            for (const auto OTD : opaqueDecls) {
+              Type interfaceType = OTD->getDeclaredInterfaceType();
```

Comment:
calling getDeclaredInterfaceType() on OpaqueTypeDecl should produce an assertion failure; are you sure you have test coverage for this?

Reply:
I used `getDeclaredInterfaceType()` because it seemed to be the most likely match for what would appear in expressions that use the type, but I'm also happy to switch it to something else more appropriate.

Reply:
I see, I missed that we set it in OpaqueResultTypeRequest::evaluate(). My apologies.

However, my original concern is that opaque return type can appear in a nested position, and we might have more than one opaque return type, eg:
```
func f() -> (some Any, some Any)? {}
```
Indeed, it appears that the declared interface type of the opaque result declaration is the entire return type of the function in this case, `(some Any, some Any)?`, so your `castTo<OpaqueTypeArchetypeType>()` will fail. You probably want to walk the declared interface type to collect all opaque archetypes that it contains.

Reply:
Ah, right. Updated this to use `transformRec` to replace any `OpaqueTypeArchetypeType`s with their `getExistentialType()`, and added a handful of tests that cover different structural situations.